Implementation of Portable::FEEvaluation::get/submit_symmetric_gradient

YiminJin opened the pull request:
This PR adds `get_symmetric_gradient()` and `submit_symmetric_gradient()` to class `Portable::FEEvaluation`.

I believe that the codes are correct. However, when I try to test them with the following program modified from step-18:
[step-18.cc.txt](https://github.com/user-attachments/files/20024898/step-18.cc.txt)
a runtime error occurs when calling `MatrixFree::cell_loop()`. The output messages are:
```
Timestep 1 at time 1
  Cycle 0:
    Number of active cells:       3712
    Number of degrees of freedom: 17226
    Assembling system rhs... norm of rhs is 3596.31
cudaDeviceSynchronize() error( cudaErrorLaunchFailure): unspecified launch failure /home/ein/build/trilinos/16.1.0/packages/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:153
Backtrace:
[0x7ec795d4c135] Kokkos::Impl::save_stacktrace()
[0x7ec795d1e27c] Kokkos::Impl::host_abort(char const*)
[0x7ec795d55a77] Kokkos::Impl::cuda_internal_error_abort(cudaError, char const*, char const*, int)
[0x7ec795d55ccf] Kokkos::Impl::cuda_device_synchronize(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
[0x7ec795d3c54d] Kokkos::Impl::ExecSpaceManager::static_fence(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
[0x7ec7cffdd31e] void Kokkos::deep_copy<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u>, double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >(Kokkos::View<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> > const&, Kokkos::View<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> > const&, std::enable_if<((is_void_v<Kokkos::ViewTraits<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >::specialize>)&&(is_void_v<Kokkos::ViewTraits<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >::specialize>))&&((((unsigned int)Kokkos::ViewTraits<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >::rank)!=(0))||(((unsigned int)Kokkos::ViewTraits<double*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >::rank)!=(0))), void>::type*)
[0x7ec7cffb5621] dealii::internal::VectorOperations::functions<double, double, dealii::MemorySpace::Default>::copy(std::shared_ptr<dealii::parallel::internal::TBBPartitioner> const&, unsigned int, dealii::MemorySpace::MemorySpaceData<double, dealii::MemorySpace::Default> const&, dealii::MemorySpace::MemorySpaceData<double, dealii::MemorySpace::Default>&)
[0x7ec7cff9f577] void dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>::copy_locally_owned_data_from<double>(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> const&)
[0x7ec7cffb2e78] dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>& dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>::operator=<double>(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> const&)
[0x7ec7cff96723] dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>::operator=(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> const&)
[0x5df973cb8b9b] void dealii::Portable::MatrixFree<3, double>::distributed_cell_loop<dealii::MatrixFreeTools::internal::ComputeDiagonalCellAction<3, 1, 2, 3, double, Step18::ElasticityOperatorQuad<3, 1> > >(dealii::MatrixFreeTools::internal::ComputeDiagonalCellAction<3, 1, 2, 3, double, Step18::ElasticityOperatorQuad<3, 1> > const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>&) const
[0x5df973ca70dc] void dealii::Portable::MatrixFree<3, double>::cell_loop<dealii::MatrixFreeTools::internal::ComputeDiagonalCellAction<3, 1, 2, 3, double, Step18::ElasticityOperatorQuad<3, 1> >, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> >(dealii::MatrixFreeTools::internal::ComputeDiagonalCellAction<3, 1, 2, 3, double, Step18::ElasticityOperatorQuad<3, 1> > const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default> const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>&) const
[0x5df973c9b6df] void dealii::MatrixFreeTools::compute_diagonal<3, 1, 2, 3, double, dealii::MemorySpace::Default, Step18::ElasticityOperatorQuad<3, 1> >(dealii::Portable::MatrixFree<3, double> const&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Default>&, Step18::ElasticityOperatorQuad<3, 1> const&, dealii::EvaluationFlags::EvaluationFlags, dealii::EvaluationFlags::EvaluationFlags, unsigned int, unsigned int, unsigned int, unsigned int)
[0x5df973c8e615] Step18::ElasticityOperator<3, 1>::compute_diagonal()
[0x5df973cafe6b] Step18::TopLevel<3, 1>::solve_linear_problem()
[0x5df973ca582f] Step18::TopLevel<3, 1>::solve_timestep()
[0x5df973c9aab8] Step18::TopLevel<3, 1>::do_initial_timestep()
[0x5df973c8e052] Step18::TopLevel<3, 1>::run()
[0x5df973c69985] main
[0x7ec795835488] 
[0x7ec79583554c] __libc_start_main
[0x5df973c683d5] _start
[BEBOP:122746] *** Process received signal ***
[BEBOP:122746] Signal: Aborted (6)
[BEBOP:122746] Signal code:  (-6)
[BEBOP:122746] [ 0] /usr/lib/libc.so.6(+0x3dcd0) [0x7ec79584bcd0]
[BEBOP:122746] [ 1] /usr/lib/libc.so.6(+0x97624) [0x7ec7958a5624]
[BEBOP:122746] [ 2] /usr/lib/libc.so.6(gsignal+0x20) [0x7ec79584bba0]
[BEBOP:122746] [ 3] /usr/lib/libc.so.6(abort+0x26) [0x7ec795833582]
[BEBOP:122746] [ 4] /opt/trilinos/16.1.0-nvcc/lib/libkokkoscore.so.4.5(_ZN6Kokkos4Impl10host_abortEPKc+0x49) [0x7ec795d1e289]
[BEBOP:122746] [ 5] /opt/trilinos/16.1.0-nvcc/lib/libkokkoscore.so.4.5(_ZN6Kokkos4Impl25cuda_internal_error_abortE9cudaErrorPKcS3_i+0x107) [0x7ec795d55a77]
[BEBOP:122746] [ 6] /opt/trilinos/16.1.0-nvcc/lib/libkokkoscore.so.4.5(_ZN6Kokkos4Impl23cuda_device_synchronizeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x1cf) [0x7ec795d55ccf]
[BEBOP:122746] [ 7] /opt/trilinos/16.1.0-nvcc/lib/libkokkoscore.so.4.5(_ZN6Kokkos4Impl16ExecSpaceManager12static_fenceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0x2d) [0x7ec795d3c54d]
[BEBOP:122746] [ 8] /home/ein/repository/dealii/build-nvcc/lib/libdeal_II.g.so.9.7.0-pre(_ZN6Kokkos9deep_copyIPdJNS_10LayoutLeftENS_6DeviceINS_4CudaENS_9CudaSpaceEEENS_12MemoryTraitsILj0EEEES1_JS2_S6_S8_EEEvRKNS_4ViewIT_JDpT0_EEERKNS9_IT1_JDpT2_EEEPNSt9enable_ifIXaaaa9is_void_vINS_10ViewTraitsISA_JSC_EE10specializeEE9is_void_vINSN_ISG_JSI_EE10specializeEEoonecvjsrSO_4rankLi0EnecvjsrSQ_4rankLi0EEvE4typeE+0x15b6) [0x7ec7cffdd31e]
[BEBOP:122746] [ 9] /home/ein/repository/dealii/build-nvcc/lib/libdeal_II.g.so.9.7.0-pre(_ZN6dealii8internal16VectorOperations9functionsIddNS_11MemorySpace7DefaultEE4copyERKSt10shared_ptrINS_8parallel8internal14TBBPartitionerEEjRKNS3_15MemorySpaceDataIdS4_EERSE_+0xac) [0x7ec7cffb5621]
[BEBOP:122746] [10] /home/ein/repository/dealii/build-nvcc/lib/libdeal_II.g.so.9.7.0-pre(_ZN6dealii13LinearAlgebra11distributed6VectorIdNS_11MemorySpace7DefaultEE28copy_locally_owned_data_fromIdEEvRKNS2_IT_S4_EE+0x1c9) [0x7ec7cff9f577]
[BEBOP:122746] [11] /home/ein/repository/dealii/build-nvcc/lib/libdeal_II.g.so.9.7.0-pre(_ZN6dealii13LinearAlgebra11distributed6VectorIdNS_11MemorySpace7DefaultEEaSIdEERS5_RKNS2_IT_S4_EE+0x3e6) [0x7ec7cffb2e78]
[BEBOP:122746] [12] /home/ein/repository/dealii/build-nvcc/lib/libdeal_II.g.so.9.7.0-pre(_ZN6dealii13LinearAlgebra11distributed6VectorIdNS_11MemorySpace7DefaultEEaSERKS5_+0x23) [0x7ec7cff96723]
[BEBOP:122746] [13] ./step-18(_ZNK6dealii8Portable10MatrixFreeILi3EdE21distributed_cell_loopINS_15MatrixFreeTools8internal25ComputeDiagonalCellActionILi3ELi1ELi2ELi3EdN6Step1822ElasticityOperatorQuadILi3ELi1EEEEEEEvRKT_RKNS_13LinearAlgebra11distributed6VectorIdNS_11MemorySpace7DefaultEEERSJ_+0xb0f) [0x5df973cb8b9b]
[BEBOP:122746] [14] ./step-18(_ZNK6dealii8Portable10MatrixFreeILi3EdE9cell_loopINS_15MatrixFreeTools8internal25ComputeDiagonalCellActionILi3ELi1ELi2ELi3EdN6Step1822ElasticityOperatorQuadILi3ELi1EEEEENS_13LinearAlgebra11distributed6VectorIdNS_11MemorySpace7DefaultEEEEEvRKT_RKT0_RSK_+0x46) [0x5df973ca70dc]
[BEBOP:122746] [15] ./step-18(_ZN6dealii15MatrixFreeTools16compute_diagonalILi3ELi1ELi2ELi3EdNS_11MemorySpace7DefaultEN6Step1822ElasticityOperatorQuadILi3ELi1EEEEEvRKNS_8Portable10MatrixFreeIXT_ET3_EERNS_13LinearAlgebra11distributed6VectorIS9_T4_EERKT5_NS_15EvaluationFlags15EvaluationFlagsESN_jjjj+0x369) [0x5df973c9b6df]
[BEBOP:122746] [16] ./step-18(_ZN6Step1818ElasticityOperatorILi3ELi1EE16compute_diagonalEv+0xdd) [0x5df973c8e615]
[BEBOP:122746] [17] ./step-18(_ZN6Step188TopLevelILi3ELi1EE20solve_linear_problemEv+0x41) [0x5df973cafe6b]
[BEBOP:122746] [18] ./step-18(_ZN6Step188TopLevelILi3ELi1EE14solve_timestepEv+0xbd) [0x5df973ca582f]
[BEBOP:122746] [19] ./step-18(_ZN6Step188TopLevelILi3ELi1EE19do_initial_timestepEv+0x206) [0x5df973c9aab8]
[BEBOP:122746] [20] ./step-18(_ZN6Step188TopLevelILi3ELi1EE3runEv+0x18) [0x5df973c8e052]
[BEBOP:122746] [21] ./step-18(main+0x6a) [0x5df973c69985]
[BEBOP:122746] [22] /usr/lib/libc.so.6(+0x27488) [0x7ec795835488]
[BEBOP:122746] [23] /usr/lib/libc.so.6(__libc_start_main+0x8c) [0x7ec79583554c]
[BEBOP:122746] [24] ./step-18(_start+0x25) [0x5df973c683d5]
[BEBOP:122746] *** End of error message ***
```
I have tried to avoid calling `MatrixFreeTools::compute_diagonal()`, and the error still occurs in `MatrixFree::cell_loop()`. 

The problem is that when I make some modifications in include/dealii/matrix_free/portable_matrix_free.template.h, it takes about an hour to re-compile with `make -j 4`. Are there any efficient methods for debugging?

tjhei commented on include/deal.II/matrix_free/portable_fe_evaluation.h:
```diff
@@ -671,6 +687,75 @@ namespace Portable
+  {
+    Assert(q_point >= 0 && q_point < n_q_points, ExcInternalError());
+    Assert(n_components_ == dim,
+           ExcMessage("Function get_symmetric_gradient() only works when the "
```
submit_symmetric_gradient

tjhei commented on include/deal.II/matrix_free/portable_fe_evaluation.h:
```diff
@@ -671,6 +687,75 @@ namespace Portable
+                      "equal."));
+
+    const gradient_type             grad = get_gradient(q_point);
+    SymmetricTensor<2, dim, Number> sym_grad;
```
Can you try if grad.symmetrize() works instead of the handwritten code you have here?

masterleinad replied:
It doesn't look like `SymmetricTensor` has `DEAL_II_HOST_DEVICE` annotations. I haven't looked if there is a reasin why we can't replace `DEAL_II_HOST` accordingly.

YiminJin replied:
I have tried it and it works. The original code is copied from `MatrixFree::get_symmetric_gradient()`, but using `symmetrize()` is neater. I have submitted a commit for this change.

tjhei commented on include/deal.II/matrix_free/portable_fe_evaluation.h:
```diff
@@ -229,6 +229,22 @@ namespace Portable
     submit_gradient(const gradient_type &grad_in, int q_point);
 
+    /**
+     * Same as above, except that the quadrature point is computed from the
```
All these comments have to be approved (not your fault as you just copied comment from above)

masterleinad commented:
> a runtime error occurs when calling `MatrixFree::cell_loop()`. The output messages are:

What does `cuda-gdb` say? Are you configuring `Kokkos` with `Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON`?

YiminJin commented:
>are you using a machine with 4 cores only or why do you use -j 4 

I have a 16-core machine, but I always use -j 4 for compiling dealii because the fan whirrs wildly when the number of processors is greater than 4. Maybe I should buy a laptop stand with fan.

>use ninja instead of make

Thanks for the advice. I will try it.

What I do not understand is why some of the other modules (such as fe, grid, numerics, etc) also have to be re-compiled when I only change include/deal.II/matrix_free/portable_matrix_free.template.h.

YiminJin commented:
> What does cuda-gdb say? Are you configuring Kokkos with Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON?

I failed to track the error with cuda-gdb. The Kokkos on my machine is not configured with Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON.

I find that the runtime error only occurs in debug mode. In release mode, there is no CUDA error, but the modified step-18 code has some problems that it does not produce the correct result. I have written a simple test code:
[test_sym_grad.cc.txt](https://github.com/user-attachments/files/20031371/test_sym_grad.cc.txt)
It also only runs in release mode, and the result is correct. Again, the CUDA error occurs in `Portable::MatrixFree::serial_cell_loop`. It seems that `Kokkos::parallel_for` was not launched properly. Could you please try the test code in DEBUG mode and help me figure out the problem?

tjhei commented:
I am happy to take a look in the next few days.

YiminJin commented:
I tried to compile trilinos with `Kokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON`, and cuda-gdb gives the same backtrace info:

```
Thread 1 "test_sym_grad" received signal SIGTRAP, Trace/breakpoint trap.
[Switching focus to CUDA kernel 0, grid 5, block (0,0,0), thread (0,0,0), device 0, sm 0, warp 2, lane 0]
0x00007fff232719f0 in void Kokkos::Impl::cuda_parallel_launch_constant_memory<Kokkos::Impl::ParallelFor<dealii::Portable::internal::ApplyKernel<2, double, LocalDeviceOperator<2, 2>, false>, Kokkos::TeamPolicy<Kokkos::Cuda>, Kokkos::Cuda> >()<<<(1,1,1),(1,128,1)>>> ()
```
The Kokkos kernel `Portable::internal::ApplyKernel` is not launched properly.

tjhei commented:
> I have written a simple test code:
> [test_sym_grad.cc.txt](https://github.com/user-attachments/files/20031371/test_sym_grad.cc.txt)

Can you please add this as a test in ``tests/matrix_free_kokkos/``?

tjhei commented:
I am pretty sure the problem is that most of SymmetricTensor is not enabled on device. I am working on a patch.

tjhei commented:
See #18486. The crash goes away with this patch for me.

YiminJin commented:
> Can you please add this as a test in tests/matrix_free_kokkos/?

Sure. However, I have a question: my outputs has both `0.000` and `-0.000`, which may change on different machines. The test code is based on `tests/matrix_free/assemble_matrix_02.cc`. That test also specifies the precision:
```deallog << std::setprecision(3);```
Yet, its output only has 0s. Could you please tell me how can I change the results to 0s?

tjhei commented:
Sure. Can you add the test with an empty .output file here?

YiminJin commented:
> Sure. Can you add the test with an empty .output file here?

Of coarse. I have added the test file to `tests/matrix_free_kokkos`. The file is renamed as `assemble_matrix_01.cc`.

tjhei commented on tests/matrix_free_kokkos/assemble_matrix_01.cc:
```diff
@@ -0,0 +1,262 @@
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+
+  FESystem<dim>   fe(FE_Q<dim>(fe_degree), dim);
```
```suggestion
  FE_Q<dim> feq(fe_degree);
  FESystem<dim>   fe(feq, dim);
```

YiminJin commented:
> Looks good after fixing the test output.

Thank you. Could you please teach me how to fix the test output? I do not know whether the standard output is like the others, i.e. only has `0`s even if I have set the precision to 3.

tjhei commented:
Hi, @YiminJin , I apologize for the long delay. I have been working towards getting things merged we need to get this PR merged. I had to wait with some of the changes before we branch of for the 9.6 release, which has now happened.
After #18534, #18719, #18486, we should be finally good to go.

tjhei commented:
Can you please rebase to the current master, @YiminJin ?

YiminJin commented:
@tjhei My apologize for the late response. I should change my mailbox on git.

I have rebased the branch. There are some failing checks, but the error information is unavailable now. I guess the test output needs to be modified. I will do that when the test results come out.

YiminJin commented:
@tjhei Could you please help me correct the test outputs? There is a failing check showing that the output of the test I added is incorrect, but I cannot find a diff file as in ASPECT.

Actually the check results are confusing: the output of the newly added test is empty, so it should cause an error in all the checkers that run the tests. But it turns out only the serial case on Jenkins throw out an error. I do not understand.

tjhei commented:
Let me know if you need help.

YiminJin commented:
@tjhei Thank you for the instruction. I updated the output and rebased again.

I find that I had a misunderstanding about the test checker: I thought that the test would fail if the result is `0.000` instead of `0`. However, when I ran `ctest -V` I found that all tests are passed. Now I see that `0` and `0.000` are the same in `numdiff`'s eyes.

tjhei commented:
Awesome, thank you! :+1:
